server/belga/io/feeding_services/spreadsheet.py

```diff
@@ -0,0 +1,218 @@
+        BelgaSpreadsheetParser().parse_titles(data[0])
+
+    def _update(self, provider, update):
+        """Load items from google spreadsheet and insert (update) to events database
+
+        If STATUS field is empty, create new item
+        If STATUS field is UPDATED, update item
+        """
+        worksheet = self._get_worksheet(provider.get('config', {}))
+
+        # Get all values to avoid reaching read limit
+        data = worksheet.get_all_values()
+        titles = [s.lower().strip() for s in data[0]]
+
+        # avoid maximum limit cols error
+        total_col = worksheet.col_count
+        if total_col < len(titles) + 3:
+            worksheet.add_cols(len(titles) + 3 - total_col)
+
+        for field in ('_STATUS', '_ERR_MESSAGE', '_GUID'):
+            if field.lower() not in titles:
+                titles.append(field)
+                worksheet.update_cell(1, len(titles), field)
+        data[0] = titles  # pass to parser uses for looking up index
+
+        parser = BelgaSpreadsheetParser()
+        items, cells_list = parser.parse(data, provider)
+        items = self._process_event_items(items, provider)
+        # add ingest item
+        yield items
+        # Update status for google sheet
+        if cells_list:
+            worksheet.update_cells(cells_list)
+
+    def _get_worksheet(self, config):
+        """Get worksheet from google spreadsheet
+
+        :return: worksheet
+        :rtype: object
+        """
+        scope = [
+            'https://spreadsheets.google.com/feeds',
+            'https://www.googleapis.com/auth/drive',
+        ]
+        url = config.get('url', '')
+        service_account = config.get('service_account', '')
+        title = config.get('worksheet_title', '')
+
+        try:
+            service_account = json.loads(service_account)
+            credentials = ServiceAccountCredentials.from_json_keyfile_dict(service_account, scope)
+            gc = gspread.authorize(credentials)
+            spreadsheet = gc.open_by_url(url)
+            permission = spreadsheet.list_permissions()[0]
+            if permission['role'] != 'writer':
+                raise IngestSpreadsheetError.SpreadsheetPermissionError()
+            worksheet = spreadsheet.worksheet(title)
+            return worksheet
+        except (json.decoder.JSONDecodeError, AttributeError, ValueError) as e:
+            # both permission and credential raise Value error
+            if e.args[0] == 15100:
+                raise IngestSpreadsheetError.SpreadsheetPermissionError()
+            raise IngestSpreadsheetError.SpreadsheetCredentialsError()
+        except gspread.exceptions.NoValidUrlKeyFound:
+            raise IngestApiError.apiNotFoundError()
+        except gspread.exceptions.WorksheetNotFound:
+            raise IngestSpreadsheetError.WorksheetNotFoundError()
+        except gspread.exceptions.APIError as e:
+            response_code = e.response.json()['error']['code']
```

Comment:
I would log also an exception's text, it's not super easy to setup all this credentials and `gspread.exceptions.APIError` has useful error text.
For example:
```
>>> e.args[0]
{
 "error": {
  "errors": [
   {
    "domain": "usageLimits",
    "reason": "accessNotConfigured",
    "message": "Access Not Configured. Drive API has not been used in project 743996421134 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/drive.googleapis.com/overview?project=743996421134 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.",
    "extendedHelp": "https://console.developers.google.com/apis/api/drive.googleapis.com/overview?project=743996421134"
   }
  ],
  "code": 403,
  "message": "Access Not Configured. Drive API has not been used in project 743996421134 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/drive.googleapis.com/overview?project=743996421134 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry."
 }
}
```